src/coreclr/vm/codeman.cpp

```diff
@@ -1381,66 +1356,92 @@ void EEJitManager::SetCpuInfo()
         CPUCompileFlags.Set(InstructionSet_LZCNT);
     }
 
-    if (((cpuFeatures & XArchIntrinsicConstants_Pclmulqdq) != 0) && CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_EnablePCLMULQDQ))
-    {
-        CPUCompileFlags.Set(InstructionSet_PCLMULQDQ);
-    }
-
     if (((cpuFeatures & XArchIntrinsicConstants_Movbe) != 0) && CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_EnableMOVBE))
     {
         CPUCompileFlags.Set(InstructionSet_MOVBE);
     }
 
-    if (((cpuFeatures & XArchIntrinsicConstants_Popcnt) != 0) && CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_EnablePOPCNT))
+    // x86-64-v4
+
+    if ((cpuFeatures & XArchIntrinsicConstants_Evex) != 0)
     {
-        CPUCompileFlags.Set(InstructionSet_POPCNT);
+        CPUCompileFlags.Set(InstructionSet_EVEX);
     }
 
-    // We need to additionally check that EXTERNAL_EnableSSE3_4 is set, as that
-    // is a prexisting config flag that controls the SSE3+ ISAs
-    if (((cpuFeatures & XArchIntrinsicConstants_Sse3) != 0) &&
-        CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_EnableSSE3) &&
-        CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_EnableSSE3_4))
+    if (((cpuFeatures & XArchIntrinsicConstants_Avx512f) != 0) &&
+        ((cpuFeatures & XArchIntrinsicConstants_Avx512f_vl) != 0)
+        ((cpuFeatures & XArchIntrinsicConstants_Avx512bw) != 0)
+        ((cpuFeatures & XArchIntrinsicConstants_Avx512bw_vl) != 0)
+        ((cpuFeatures & XArchIntrinsicConstants_Avx512cd) != 0)
+        ((cpuFeatures & XArchIntrinsicConstants_Avx512cd_vl) != 0)
+        ((cpuFeatures & XArchIntrinsicConstants_Avx512dq) != 0)
+        ((cpuFeatures & XArchIntrinsicConstants_Avx512dq_vl) != 0))
+    {
```

Comment:
It might require a R2R breaking change, but it would be nice if we could just compress these down to 1 bit and cleanup the `CorInfoInstructionSet` handling a bit as well.

We're never going to support, in the JIT at least, having the baseline AVX512 ISAs independent, so it'd be nicer if we just had `DOTNET_EnableAVX512`, `XArchIntrinsicConstants_Avx512`, and `InstructionSet_Avx512` instead. That would also allow us to do simpler queries in the JIT around ISA support

At the same time, the nested classes are implied by other information. For example, the `X64`, `VL`, and `V512` nested classes are all implied by the enclosing type being supported and another implicit bit of information such as the target architecture or if the base class is supported. So we really don't "need" all the extra logic and instruction set IDs to support those.